### ISSUE TITLE
Spire agent SDS does not respond with SPIFFECertValidatorConfig occasionally

### ISSUE THREAD

**Issue by grameshtwilio** [spiffe/spire#4608]
### Setup
 [Istio Spire integration sample](https://github.com/istio/istio/tree/1.20.0-beta.0/samples/security/spire#integrating-spire-as-a-ca-through-envoys-sds-api) on a KinD cluster

### Observed
When a pod such as the Istio Ingress gateway is deployed after the Spire agent is already present on a node (this is usually the case), the ROOTCA's ValidationContext is **not** `SPIFFECertValidatorConfig`

```
 curl http://localhost:15000/config_dump | jq '.configs | last | .dynamic_active_secrets | last | .secret.validation_context'

{
  "trusted_ca": {
    "inline_bytes": "<redacted>"
  }
}

```

### Expected
The ROOTCA's ValidationContext should always be `SPIFFECertValidatorConfig`. Config dump on the Ingress gateway pod

```
curl http://localhost:15000/config_dump | jq '.configs | last | .dynamic_active_secrets | last | validation_context'

{
  "trusted_ca": {},
  "custom_validator_config": {
    "name": "envoy.tls.cert_validator.spiffe",
    "typed_config": {
      "@type": "type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.SPIFFECertValidatorConfig",
      "trust_domains": [
        {
          "name": "example.org",
          "trust_bundle": {
            "inline_bytes": "<redacted>"
          }
        }
      ]
    }
  }
}

````
By bouncing the Spire agent, this issue resolves majority of the time.

### Hypothesis
1. When a new pod requests for a workload identity from the Spire agent, it has not been registered with the Spire server. During this period, Envoy requests both the workload and CA certificate. 
2. The Spire agent responds with the CA cert with the correct ValidationContext but not the workload certificate as witnessed in the agent logs ```time="2023-10-25T19:37:17Z" level=error msg="Error building stream secrets response" error="rpc error: code = InvalidArgument desc = workload is not authorized for the requested identities [\"default\"]" method=StreamSecrets pid=11635 service=SDS.v3 subsystem_name=endpoints```
3. Envoy retries until the workload certificate is available
4. The Spire agent responds with both the workload and CA certificate but this time ValidationContext is incorrect because the subsequent requests from envoy do not include the [Node](https://github.com/envoyproxy/go-control-plane/blob/v0.11.1/envoy/service/discovery/v3/discovery.pb.go#L165) property, and this ends up evaluating [supportsSPIFFEAuthExtension](https://github.com/spiffe/spire/blob/v1.7.2/pkg/agent/endpoints/sdsv3/handler.go#L507) to `false`
5. When the spire agent is restarted, it responds with the correct CA cert ValidationContext because the workload pod is already registered.

* **Version**: 1.7.2, 1.5.4
* **Platform**: Kubernetes 1.23, 1.24
* **Subsystem**: Spire agent

**Comment by azdagron** [comment]
So you're hypothesis is that Envoy only populates `Node` on the first request on the stream?

**Comment by grameshtwilio** [comment]
Envoy Version 1.25.4. 

I tested this with a patch locally and it seemed to work.

https://github.com/spiffe/spire/compare/main...grameshtwilio:gramesh/sds-patch?expand=1

Not sure if it's the right approach

**Comment by azdagron** [comment]
I have a few nitpicks but otherwise that seems like a fine, robust approach. If you wouldn't mind opening a PR, that would be great.

**Comment by grameshtwilio** [comment]
Thanks @azdagron. PR https://github.com/spiffe/spire/pull/4611

**Comment by azdagron** [comment]
Fixed by #4611.
